Add a BufferGeometry based Subdivision Modifier

centerionware opened the pull request:
This started from the original subdivision modifier, and works in the
same way except it has different output - and is much MUCH faster and
far easier on RAM. (From >1m and >3gb (queue crashing browsers) to <10
seconds and <500mb on my tests).

Expected input: Index'd Buffer Geometry or THREE.Geometry - with uv's.
Output: Unindex'd Buffer Geometry.

Also introduced is a class I call the 'TypedArrayHelper' which provides
THREE objects (Such as Face, Vec3, Vec2) as 'registers' (I was thinking
like a CPU register), and controls the typed arrays underneath. It
automatically resizes them and provides a trim() function. While
resizing arrays is slow, it's still all way faster than millions of vec3
objects on a regular Geometry.

centerionware committed:
Add a BufferGeometry based Subdivision Modifier

This started from the original subdivision modifier, and works in the
same way except it has different output - and is much MUCH faster and
far easier on RAM. (From >1m and >3gb (queue crashing browsers) to <10
seconds and <500mb on my tests).

Expected input: Index'd Buffer Geometry or THREE.Geometry - with uv's.
Output: Unindex'd Buffer Geometry.

Also introduced is a class I call the 'TypedArrayHelper' which provides
THREE objects (Such as Face, Vec3, Vec2) as 'registers' (I was thinking
like a CPU register), and controls the typed arrays underneath. It
automatically resizes them and provides a trim() function. While
resizing arrays is slow, it's still all way faster than millions of vec3
objects on a regular Geometry.

mrdoob commented:
Interesting. How come you went with indexed BufferGeometry as input instead of unindexed BufferGeometry?

centerionware commented:
Because it needs to build edge lists, and index'd is much easier to do that with as there are no duplicated vertices.

centerionware commented on examples/js/modifiers/BufferSubdivisionModifier.js:
```diff
@@ -0,0 +1,753 @@
+var TypedArrayHelper = function (size, registers, register_type, array_type, unit_size, accessors) {
+    this.array_type = array_type;
+    this.register_type = register_type;
+    this.unit_size = unit_size;
```
I just realized unit_size can be set from accessors.length instead.

centerionware replied:
But that may not always be ideal so I will leave it. Eg: when using vec4 registers on a vec3 typed array for some reason I can't think of right now.

centerionware committed:
Various Fixes

Fixed normal generation according to
https://www.opengl.org/wiki/Calculating_a_Surface_Normal
maybe various other cleanups.

centerionware committed:
Revert "Various Fixes"

This reverts commit 4348a5f60b85ad8381e7ed3f19c0602a23bb87c4.

centerionware committed:
Merge remote-tracking branch 'refs/remotes/mrdoob/dev' into dev

centerionware committed:
Revert "Revert "Various Fixes""

This reverts commit c972529a2e0d869ac40d29409beb21420a833ad2.

centerionware commented:
wow, so I had some trouble getting a patch for this in last night. The fixes are for normal creation. so now there should be a 'Revert Revert Various Fixes' commit.. *facepalm*

mrdoob commented:
Do you mind updating `webgl_modifier_subdivision` so it uses this instead? If the output is 1:1 we could probably remove the old one.

centerionware commented:
The original subdivision modifier returned the original geometry object that was passed in, however this version does not do that. Because the original geometry can be either a three.geometry or index'd three.geometry it returns a new geometry object - in this case, an unindex'd buffer geometry (Only because I couldn't get an index'd buffer geometry to render). So the existing example doesn't really want to work, as it shares geometry objects with the mesh multi material object. I suppose I could make it fill in the data in the incoming object, that would allow the example script to work with it, but would add even more overhead (So maybe with a bool at the constructor)?

centerionware commented:
or I could simply rewrite the example to work with it. I'll leave the choice up to you and do it whichever way you think is best.

mrdoob commented:
> or I could simply rewrite the example to work with it. 

I think that works 😊

tschw commented on examples/js/modifiers/BufferSubdivisionModifier.js:
```diff
@@ -0,0 +1,766 @@
+    resize: function (new_size) {
+        if (new_size == 0) new_size = 8;
+        var nBuffer = new this.array_type(new_size * this.unit_size);
+        nBuffer.set(this.buffer.slice(0, this.length * this.unit_size));
```
Should be using `subarray` here to use a view instead of a (copied) `slice`. Halves the copying and gets you rid of a major allocation in favor of a small one. In case `this.buffer.length` < `nBuffer.length` (which I guess might be the case) may just call `set` directly, even.

centerionware replied:
I want to look into this, anything to reduce allocations and copies is a great thing, especially when working with extremely large meshes (I'm rendering some with > 800k verts, which means arrays of > 2.4m floats, with uv's and normals for each vert too)
If I don't get to it before this is all sync'd (I'm a lot busier in life these days) I hope someone will look at this idea. *poke*

centerionware replied:
this resize function though is almost always only used for growing the existing buffer automatically (when pushing elements). It is however used in one case when almost always (>99% of the time) nBuffer.length is < this.buffer.length, which is in the trim_size() function which is to always be called once the work is done and the buffer is filled, so the actual buffer length isn't larger than what's expected.

tschw replied:
[TypedArray.prototype.slice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/slice) does not appear to be implemented in all browsers. So you should probably address this issue before merge.

And as said, also needlessly doubles your peak memory consumption. Doesn't matter how often it's run, may still be just enough to hit your browser's limit on a large mesh.

centerionware replied:
I can say for absolute sure that page is wrong. Chrome and MS Edge browser absolutely support it, I don't know about Opera as I don't run that. However, I have made the changes, and still, when growing an array it's going to still double the peak usage - there is no way (That I know of!) to avoid it because one cannot simply grow a typed array(that I know of!). To grow it, a new array needs to be allocated that's bigger than the old, then filled with the old data (even if that's via subarray)

tschw replied:
> I can say for absolute sure that page is wrong.

Outdated, probably - good to know the situation is improving.

> one cannot simply grow a typed array(that I know of!)

No. Typed arrays have a fixed size, so you definitely have to reallocate and copy it.

Maybe I was unclear?

Using `slice` you copy it *twice* instead of just *once* and also need twice the amount of memory. Just use `set`, unless the (old) source array can actually be larger than the (new) destination (I guess it doesn't happen?), in that case you'd have to create a view that fits into the new array using `subarray` (this one does not copy - it creates a view on the same buffer) and pass that to `set`.

Making sense now?

centerionware replied:
Okay makes a lot more sense now. This final commit should do it, and checks to ensure the sizes are all right. It is actually possible for the old source to be larger than the new destination, I haven't yet looked at why this may be, but it can be. A simple if and it should be all good now.
Thanks for the input! if you spot anything else that can be optimized, please let me know! This  seems quite a bit faster now rendering suzanne and walt head with 3 subdivs than even yesterday.

centerionware committed:
change the example to use BufferSubdivisionModifier

It's almost all working except for LatheGeometry. I'm not entirely sure
what is different about the lathegeometry , but I will investigate later
unless someone else figures it out first.

centerionware commented:
I put this in the last commit details, but I couldn't figure out the difference between the lathe geometry and other geometry so it doesn't subdivide like the others. Everything else is working.

centerionware commented:
I honestly don't remember doing that. i must have had an older copy.  same goes for changing it from phong to lambert.

centerionware committed:
change back to v2 and use phongmaterial

change back to v2 and use phongmaterial

centerionware commented:
changing those back to vector2's and meshphongmaterial did the trick! it's all working now.

WestLangley commented:
Hmmm... You are not disposing of your geometries properly. Check `renderer.info()`; It should be an easy fix.

centerionware committed:
dispose of geometries

see title.

WestLangley commented on examples/webgl_modifier_subdivision.html:
```diff
@@ -173,22 +173,28 @@
 
-				if ( cube ) {
-
+				if ( cube != false ) {
```
Isn't cube an `object3D`? Do you mean `undefined`? I think the code you inherited had unusual variable names. : - )

centerionware replied:
well, yeah. It did. But I set it to false in the init() function so this would work.

WestLangley commented on examples/webgl_modifier_subdivision.html:
```diff
@@ -173,22 +173,28 @@
 					scene.remove( group );
-					scene.remove( cube );
+					scene.remove(cube);
+					scene.remove(mesh);
```
Isn't `mesh` a child of `group`?

centerionware replied:
I think so, it does have a group.add(mesh) in there. was mostly poking in the dark trying to find the thing that was leaking, it was when I did geometry.dispose() that things stopped leaking, so I think that's all that was really required. I don't think the other stuff hurts though, except for making it uglier and longer than necessary.

centerionware commented on examples/webgl_modifier_subdivision.html:
```diff
@@ -205,51 +211,40 @@
-				smooth.mergeVertices();
-				smooth.computeFaceNormals();
-				smooth.computeVertexNormals();
+				// mergeVertices(); is run in the subdivision modifier 
```
also, the normals are computed post subdivision by the subdiv mod, so no need to run that now.

centerionware committed:
Remove typedarray.slice's

in favor of typedarray.subarray's.

centerionware committed:
slight cleanup to subarray

simply set the new buffers contents to the old buffers contents if
possible, otherwise use subarrays

tschw commented:
> It is actually possible for the old source to be larger than the new destination, I haven't yet looked at why this may be, but it can be.

Maybe using it to cut off the spare end of the buffer when you're done?

> This seems quite a bit faster now rendering suzanne and walt head with 3 subdivs than even yesterday.

Figured it might turn out significant after what you wrote in the description.

I'm wondering whether it would make sense to have a factorization for geometry access (beyond the scope of this PR) that provides unified access to the various encapsulations (individually buffered, interleaved, legacy geometry), takes care of memory management and an (optionally) optional index buffer and supports any number of vertex attributes. This would make all conversion between those variants trivial and algorithms like subdivision, extrusion, tessellation, CSG ... could use it for I/O and wouldn't have to worry about the encoding. It would also provide a utility for linear interpolation including all attributes (since pretty much all algorithms need that one).

mrdoob commented:
Thanks!

WestLangley commented:
> @input THREE.Geometry, or index'd THREE.BufferGeometry with faceUV's (Not vertex uv's)

Is this modifier intended to support indexed `BufferGeometry` -- such as `SphereBufferGeometry`?

centerionware commented:
the function THREE.BufferSubdivisionModifier.prototype.modify = function (geometry) { ... } at line 378 checks to ensure it's either an instaceof THREE.Geometry or THREE.BufferGeometry. If there are other types of indexed geometry that I'm unaware of they can be added in there. It doesn't really check to ensure it's indexed, but it does require them to be.
So, yes.

WestLangley commented:
I get errors if I change `SphereGeometry` to `SphereBufferGeometry` in the example. Do you mind checking?

>It doesn't really check to ensure it's indexed, but it does require them to be.

It should check, then. It would be sufficient to check if the `index` property is not `null`.

>with faceUV's (Not vertex uv's)

Can you explain what you mean by that?

centerionware commented:
I meant normals, not UV's. dunno how I didn't catch that reading it over and over. 

I'll take a look at the SphereGeometry vs SphereBufferGeometry.
It would be a good idea to perform that check and at least warn in the console.log, you're right.

centerionware commented:
updated previous comment, dunno if github will notify you of that.

centerionware commented:
Yeah, the more I think about it, the more I realize even with indexed the regular way of doing it would be to still duplicate the vertex if the uv's aren't the same. This was bad thinking on my part, but allowed me to reduce the # of vertices I needed to store in memory in my own project that I created it for initially when 100% of the uv's were never the same (Thus eliminating the memory benefit of using indexed in the first place). I'll go ahead and add a boolean that can be set to use "face uv's" or proper vertex uv's sometime tomorrow that will default to use vertex uv's.

rfm1201 commented:
I tried to understand why the example page doesn’t work with buffered geometry.
I only try to understand with the simpliest one, aka BoxBufferGeometry.
My conclusion is that the problem is not in the subdivision code but in the calculateVertexCount method of BoxBufferGeometry that return 96 (in the case of the test page).
All the arrays are allocated using a multiple of that vertexCount.
In fact the real number should be 54 (9 vertex by side, 6 sides).
As unIndexIndexedGeometry of TypedArrayHelper use the face count to allocate it’s own buffer, when it try to put all the old array in the new one there is a problem because the old one is far too big.

As the other buffered geometry seems to have the same behavior, I suppose the problem is the same in the other buffered geometries.

rfm1201 commented:
I check my own version of BoxBufferGeometry and there is still a problem with UVs.

centerionware commented:
@rfm1201 yes, UV's will be an issue until I go ahead and make a patch. right now it expects UV's to be per index, not per vertex. Every time an index is used to specify a vertex, a different UV is used. This was because the application I wrote it for never had the same UV's, but very - very often shared the same vertex so I had merged them to make things slightly quicker and more memory efficient when generating the terrain I was dealing with. I do plan on fixing it by adding a bool that defaults to using per-vertex UV's instead. Just haven't had the time today to do it, and probably won't have the time until Saturday.

WestLangley commented:
>right now it expects UV's to be per index, not per vertex. 

`Geometry` has per-face UV's. `BufferGeometry` has per-vertex UV's. Per-index UV's do not exist in three.js.

centerionware commented:
When I say "per face uv's" I mean per-index uv's.. Geometry has "face UVs"  but really they're per-index UV's, there's 3 UV's per face - and each face consists of 3 indices (Tris, not quads, or there would be 4 UV's per face). I meant to make a patch today but have been very busy again. I hope to be able to get to it soon.

centerionware commented:
There's another issue, my SphereBufferGeometry is not indexed (geometry.index is coming up null), so it's not working even after making a patch.

centerionware commented:
Hmm, It seems that although the Buffer versions of the geometries are indexed, they're still using the same properties as unindexed (EG: Each index is referencing a different vertex - even if the vertex is identical to one that exists) and there is (as far as I know) no mergeVertices function like there is with THREE.Geometry, so it's tearing all the meshes.
